remote_engine_client/src/cached_router.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! Cached router
+
+use std::num::NonZeroUsize;
+
+use ceresdbproto::storage;
+use clru::CLruCache;
+use router::RouterRef;
+use snafu::{OptionExt, ResultExt};
+use table_engine::remote::model::TableIdentifier;
+use tokio::sync::Mutex;
+use tonic::transport::Channel;
+
+use crate::{channel::ChannelPool, config::Config, error::*};
+
+pub struct CachedRouter {
+    /// Router which can route table to its endpoint
+    router: RouterRef,
+
+    /// Cache mapping table to channel of its endpoint
+    // TODO: should be replaced with a cache(like "moka")
+    // or partition the lock.
+    cache: Mutex<CLruCache<TableIdentifier, Channel>>,
+
+    /// Channel pool
+    channel_pool: ChannelPool,
+}
+
+impl CachedRouter {
+    pub fn new(router: RouterRef, config: Config) -> Self {
+        Self {
+            router,
+            cache: Mutex::new(CLruCache::new(
+                NonZeroUsize::new(config.route_cache_size).unwrap(),
+            )),
+            channel_pool: ChannelPool::new(config),
+        }
+    }
+
+    pub async fn route(&self, table_ident: &TableIdentifier) -> Result<Channel> {
+        // Find in cache first.
+        let channel_opt = {
+            let mut cache = self.cache.lock().await;
+            cache.get(table_ident).cloned()
+        };
+
+        let channel = if let Some(channel) = channel_opt {
+            // If found, return it.
+            channel
+        } else {
+            // If not found, do real route and put it into cache, and return then.
+            let mut cache = self.cache.lock().await;
+
+            // Double check here, it may have been put by others.
+            let channel_opt = cache.get(table_ident).cloned();
+            if let Some(channel) = channel_opt {
+                channel
+            } else {
+                self.do_route(table_ident).await?
```

Comment:
It seems the new channel has not been updated into the cache?

Reply:
Fix.